### ISSUE TITLE
Add insights about why a variant was selected

### ISSUE THREAD

**Issue by melix** [gradle/gradle#8786]
### Context

With [attribute based matching](https://docs.gradle.org/current/userguide/dependency_management_attribute_based_matching.html), it is possible that a build passes but you still have to understand _why_. For example, a variant is selected but you didn't expect it to be. In those cases, we don't have good insights about why this happened:

- which variants were candidates
- what attributes were considered during matching
- what attributes were compatible
- what was preferred by a compatibility rule (which one?)

We have good insights when selection _fails_, but not when it _works_.

In short, it would be nice to have diagnostics about how the selection process of a variant works. This should probably also appear in [build scans](https://scans.gradle.com/)
